src/main/java/org/opensearch/security/dlic/rest/validation/CredentialsValidator.java

```diff
@@ -99,4 +111,53 @@ public boolean validate() {
         }
         return true;
     }
+
+    private boolean regexBasedPasswordValidation(final String regex, final String password) {
+        // Password can be null for an existing user. Regex will validate password if present
+        if (!Pattern.compile("^" + regex + "$").matcher(password).matches()) {
+            if (log.isDebugEnabled()) {
+                log.debug("Regex does not match password");
+            }
+            this.errorType = ErrorType.INVALID_PASSWORD;
+            return false;
+        }
+        return true;
+    }
+
+    private boolean scoreBasedPasswordValidation(final PasswordValidationScoreStrength scoreStrength,
+                                                 final String username, final String password) {
+        final int minPasswordLength = opensearchSettings.getAsInt(SECURITY_RESTAPI_PASSWORD_SCORE_BASED_VALIDATION_MIN_PASSWORD_LENGTH, 8);
+        if (password.length() < minPasswordLength) {
+            this.errorType = ErrorType.INVALID_PASSWORD;
+            if (log.isDebugEnabled()) {
+                log.debug("Password is too small");
+            }
+            return false;
+        }
+        final Strength strength = new Zxcvbn().measure(password, ImmutableList.of(username));
+        if (strength.getScore() < scoreStrength.score()) {
+            this.errorType = ErrorType.INVALID_PASSWORD;
+            if (log.isDebugEnabled()) {
+                log.debug("Password is too weak");
+            }
+            return false;
+        }
+        final boolean checkUsernameSimilarity = this.opensearchSettings.getAsBoolean(
+                SECURITY_RESTAPI_PASSWORD_SCORE_BASED_VALIDATION_CHECK_USERNAME_SIMILARITY, true);
+        if (checkUsernameSimilarity) {
+            final boolean similar = strength.getSequence()
+                    .stream()
+                    .filter(m -> m.matchedWord != null)
```

Comment:
Attach @DarshitChanpura 's question here:

>What is the behavior when username and password is only 1 or 2 letters off? Will the password be allowed as valid?

I can see that we are handling the case of 100% matching. However, what if I'm having a username called "Adminacc" and I want to set up my password as "Admin" (Let's discard the length validation here)?  In other words, do we need to set up a scope for similarity?

Reply:
I made a mistake here which i fixed. The scope in this case is just a username or a list of user names which will be involved into scoring.